src/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs

```diff
@@ -48,41 +48,88 @@ public async Task<IEnumerable<EmissionsForecast>> GetCurrentForecastDataAsync(ID
     {
         using (var activity = Activity.StartActivity())
         {
-            TimeSpan windowSize = GetDurationOrDefault(props);
             _logger.LogInformation("Aggregator getting carbon intensity forecast from data source");
 
             var forecasts = new List<EmissionsForecast>();
             foreach (var location in GetLocationOrThrow(props))
             {
                 var forecast = await this._dataSource.GetCurrentCarbonIntensityForecastAsync(location);
-                var firstDataPoint = forecast.ForecastData.First();
-                var lastDataPoint = forecast.ForecastData.Last();
-                forecast.DataStartAt = GetOffsetOrDefault(props, CarbonAwareConstants.Start, firstDataPoint.Time);
-                forecast.DataEndAt = GetOffsetOrDefault(props, CarbonAwareConstants.End, lastDataPoint.Time + lastDataPoint.Duration);
-                forecast.Validate();
-                forecast.ForecastData = IntervalHelper.FilterByDuration(forecast.ForecastData, forecast.DataStartAt, forecast.DataEndAt);
-                forecast.ForecastData = forecast.ForecastData.RollingAverage(windowSize);
-                forecast.OptimalDataPoint = GetOptimalEmissions(forecast.ForecastData);
-                if (forecast.ForecastData.Any())
-                {
-                    forecast.WindowSize = forecast.ForecastData.First().Duration;
-                }
-                forecasts.Add(forecast);
+                var emissionsForecast = ProcessAndValidateForecast(forecast, props);
+                forecasts.Add(emissionsForecast);
             }
 
             return forecasts;
         }
     }
 
+    /// <inheritdoc />
+    public async Task<EmissionsForecast> GetForecastDataAsync(IDictionary props)
+    {
+        EmissionsForecast forecast;
+        using (var activity = Activity.StartActivity())
+        {
+            ValidateInput(props);
+          
+            var locations = (IEnumerable<Location>) props[CarbonAwareConstants.Locations]!;
+            var forecastRequestedAt = GetOffsetOrDefault(props, CarbonAwareConstants.ForecastRequestedAt, default);
+            _logger.LogDebug($"Aggregator getting carbon intensity forecast from data source for location {locations.First()} and requestedAt {forecastRequestedAt}");
+
+            forecast = await this._dataSource.GetCarbonIntensityForecastAsync(locations.First(), forecastRequestedAt);
+            var emissionsForecast = ProcessAndValidateForecast(forecast, props);
+            return emissionsForecast;
+        }
+  
+    }
+    private void ValidateInput(IDictionary props)
```

Comment:
nit: would have preferred a `ValidatateForecastInputs` name or similar, since the method is not generic validator based on a set of "Expected Props" vs. "Actually Found Props"...

Reply:
Hi @dnastrain . I think it makes sense, but seems like we are not very pro-active on using framework specific classes that are designed to that type of validation. (I.e `ValidationAttribute`)